doc/read-the-docs-site/extensions-flags-pragmas.rst

```diff
@@ -0,0 +1,83 @@
+
+
+Extensions
+-------------------------------------------------
+
+Plutus Tx modules should use the ``Strict`` extension: ::
+
+  {-# LANGUAGE Strict #-}
+
+Unlike in Haskell, function applications in Plutus Tx are strict.
+In other words, when evaluating ``(\x -> 42) (3 + 4)`` the expression ``3 + 4`` is evaluated first, before evaluating the function body (``42``), even though ``x`` is not used in the function body.
+The ``Strict`` extension ensures that let bindings are also (by default) strict, for instance, evaluating
+``let x = 3 + 4 in 42`` evaluates ``3 + 4`` first, even though ``x`` is not used.
+
+Bang patterns and lazy patterns can be used to explicitly specify whether a let binding is strict or non-strict, as in ``let !x = 3 + 4 in 42`` (strict) and ``let ~x = 3 + 4 in 42`` (non-strict).
+At this time, it is not possible to make function applications non-strict: ``(\(~x) -> 42) (3 + 4)`` still evaluates ``3 + 4`` strictly.
+
+Making let bindings strict by default has the following advantages:
+
+* It makes let bindings and function applications semantically equivalent, e.g., ``let x = 3 + 4 in 42`` has the same semantics as ``(\x -> 42) (3 + 4)``.
+  This is what one would come to expect, as it is the case in most other programming languages, regardless of whether the language is strict or non-strict.
+* Untyped Plutus Core programs, which are compiled from Plutus Tx, are not evaluated lazily (unlike Haskell), that is, there is no memoization of the results of evaluated expressions.
+  Thus using non-strict bindings can cause an expression to be inadvertently evaluated for an unbounded number of times.
+  Consider ``let x = <expensive> in \y -> x + y``.
+  If ``x`` is non-strict, ``<expensive>`` will be evalutated every time ``\y -> x + y`` is applied to an argument, which means it can be evaluated 0 time, 1 time, 2 times, or any number of times (this is not the case if lazy evaluation was employed).
+  On the other hand, if ``x`` is strict, it is always evaluated once, which is at most one more time than what is necessary.
+
+Flags
+-------------------------------------------------
+
+GHC has a variety of optimization flags, many of which are on by default.
+Although Plutus Tx is, syntactically, a subset of Haskell, it has different semantics and a different evaluation strategy (Haskell: non-strict semantics, call by need; Plutus Tx: strict semantics, call by value).
+As a result, some GHC optimizations are not helpful for Plutus Tx programs, and can even be harmful, in the sense that it can make Plutus Tx programs less efficient, or fail to be compiled.
+An example is the full laziness optimization, controlled by GHC flag ``-ffull-laziness``, which floats let bindings out of lambdas whenever possible.
+Since Untyped Plutus Core does not employ lazy evaluation, the full laziness optimization is never beneficial, and can sometimes make a Plutus Tx program more expensive.
```

Comment:
> the full laziness optimization is never beneficial

How come? What about transforming, say,

```haskell
map (\x -> let y = <expensive> in x + y) xs
```

into

```haskell
let y = <expensive> in map (\x -> x + y) xs
```

assuming the `let`  is non-strict and  `xs` is non-empty?

Reply:
Unless I'm missing something, `<expensive>` is evaluated the same number of times in both cases. Moving a non-strict binding never matters.

Reply:
Sorry, I said the wrong thing. I mean, assuming `let` is strict, of course.

Reply:
I doubt `-ffull-laziness` moves strict bindings; it'd need to prove `xs` is non-empty. And it's called full _laziness_ after all. I guess it might be possible. But turning it off makes the test cases we have strictly better.

Reply:
I think there is a misunderstanding here. Floating out bindings can be helpful - indeed, we have our own FloatOut pass. The problem with `full-laziness` is that specifically in GHC it does the reverse binder-swap transformation, which can lead to us re-evaluating things. _That_ is the thing that is bad for us. If it didn't do that, it would probably be fine and maybe even beneficial.

Reply:
> The problem with full-laziness is that specifically in GHC it does the reverse binder-swap transformation, which can lead to us re-evaluating things.

Understood, but I don't think we need to explain this detail here. Should suffice to say that `full-laziness` can be bad. I'll change "never beneficial" to "rarely beneficial".